system/lib/emmalloc.c

```diff
@@ -264,49 +264,49 @@ static bool region_is_in_use(Region *r)
   return r->size == region_ceiling_size(r);
 }
 
-static uint32_t size_of_region_from_ceiling(Region *r)
+static size_t size_of_region_from_ceiling(Region *r)
 {
-  uint32_t size = region_ceiling_size(r);
+  size_t size = region_ceiling_size(r);
   return DECODE_CEILING_SIZE(size);
 }
 
 static bool debug_region_is_consistent(Region *r)
 {
   assert(r);
-  uint32_t sizeAtBottom = r->size;
-  uint32_t sizeAtCeiling = size_of_region_from_ceiling(r);
+  size_t sizeAtBottom = r->size;
+  size_t sizeAtCeiling = size_of_region_from_ceiling(r);
   return sizeAtBottom == sizeAtCeiling;
 }
 
 static uint8_t *region_payload_start_ptr(Region *region)
 {
-  return (uint8_t*)region + sizeof(uint32_t);
+  return (uint8_t*)region + sizeof(size_t);
 }
 
 static uint8_t *region_payload_end_ptr(Region *region)
 {
-  return (uint8_t*)region + region->size - sizeof(uint32_t);
+  return (uint8_t*)region + region->size - sizeof(size_t);
 }
 
-static void create_used_region(void *ptr, uint32_t size)
+static void create_used_region(void *ptr, size_t size)
 {
   assert(ptr);
-  assert(HAS_ALIGNMENT(ptr, sizeof(uint32_t)));
-  assert(HAS_ALIGNMENT(size, sizeof(uint32_t)));
+  assert(HAS_ALIGNMENT(ptr, sizeof(size_t)));
+  assert(HAS_ALIGNMENT(size, sizeof(size_t)));
   assert(size >= sizeof(Region));
-  *(uint32_t*)ptr = size;
-  ((uint32_t*)ptr)[(size>>2)-1] = size;
+  *(size_t*)ptr = size;
+  ((size_t*)ptr)[(size/sizeof(size_t))-1] = size;
```

Comment:
Can you instead make this a macro with a 32-bit vs 64-bit `>>2` vs `>>3` behavior? Sure, compilers will certainly optimize a division, but they won't do it in -O0, and e.g. reading disassemblies when running a debug build of emmalloc to detect issues will then use different code.